sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -1466,10 +1473,12 @@ object CodeGenerator extends Logging {
           val startTime = System.nanoTime()
           val result = doCompile(code)
           val endTime = System.nanoTime()
-          def timeMs: Double = (endTime - startTime).toDouble / NANOS_PER_MILLIS
+          val duration = endTime - startTime
+          val timeMs: Double = duration.toDouble / NANOS_PER_MILLIS
           CodegenMetrics.METRIC_SOURCE_CODE_SIZE.update(code.body.length)
           CodegenMetrics.METRIC_COMPILATION_TIME.update(timeMs.toLong)
           logInfo(s"Code generated in $timeMs ms")
+          _compileTime.add(duration)
```

Comment:
I think `timeMs` is good enough to be recorded.

Reply:
`timeMs` lost a lot of precision. Because `SQLQueryTestSuite` contains 231 test cases contain a lot of SQL, tens of thousands of SQL are executed. The loss of accuracy will increase.